drivers/clock_control/Kconfig

```diff
@@ -9,6 +9,14 @@
 #
 # Clock controller drivers
 #
+
+config CLOCK_CONTROL_HAS_SUBSYS_DRIVER
```

Comment:
I think we want a different name for this option.

It's especially the 'SUBSYS' part that I'm concnerned about. I understand that in this particular case, an out-of-tree subsystem needed to force a particular driver.

But it might be in the future that an out-of-tree driver, or out-of-tree app might want to force their driver. So the 'subsys' term doesn't work. Perhaps 

CLOCK_CONTROL_HAS_OOT_DRIVER
CLOCK_CONTROL_HAS_FORCED_DRIVER
CLOCK_CONTROL_HAS_OUTSIDE_DRIVER

I'd rather not use the term 'external' as it is not as precise as out-of-tree.

Reply:
I agree, the name is not the best. I don't mind long names here, so something `CLOCK_CONTROL_HAS_OUT_OF_TREE_DRIVER` is fine by me. Other ideas/suggestions are welcome as well.

Reply:
I assume "CLOCK_CONTROL_HAS_SUBSYS_DRIVER" and the "if  CLOCK_CONTROL_HAS_SUBSYS_DRIVER"  guard was introduced to allow the OOT driver and the Zephyr driver to use the same Kconfig symbols without interfering each other.

In my view the usage of the same Kconfig symbols should be controlled by the driver. The Zephyr driver or the OOT driver should enable the Kconfig symbols by a non user selectable Kconfig option (e.g. CLOCK_CONTROL_NRF_HAS_DRIVER).

A driver that wants to activate the options just has to select CLOCK_CONTROL_xxx_HAS_DRVER.

```
menuconfig CLOCK_CONTROL_MY_NRF
	bool "My NRF Clock controller support"
	depends on SOC_COMPATIBLE_NRF
        select CLOCK_CONTROL_NRF_HAS_DRIVER
	help
	  Enable my specific support for the Nordic Semiconductor nRFxx series SoC clock driver.
```

```
menuconfig CLOCK_CONTROL_NRF
	bool "NRF Clock controller support"
	depends on SOC_COMPATIBLE_NRF
        select CLOCK_CONTROL_NRF_HAS_DRIVER
	help
	  Enable support for the Nordic Semiconductor nRFxx series SoC clock
	  driver.

if CLOCK_CONTROL_NRF_HAS_DRIVER

config CLOCK_CONTROL_NRF_IRQ_PRIORITY
	int "Power Clock Interrupt Priority"
	range 0 7
	default 1
	help
	  The interrupt priority for Power Clock interrupt.

  ...
```


That would also allow to keep the Kconfig option for the driver mostly in one place.

Reply:
> I assume "CLOCK_CONTROL_HAS_SUBSYS_DRIVER" and the "if CLOCK_CONTROL_HAS_SUBSYS_DRIVER" guard was introduced to allow the OOT driver and the Zephyr driver to use the same Kconfig symbols without interfering each other.

It should be `if !CLOCK_CONTROL_HAS_SUBSYS_DRIVER`.
This is not the reason. The driving use-case for this is for out-of-tree subsystems (or anything, really) to have a way of de-selecting in-tree drivers. This could be an external subsystem that is Zephyr agnostic and needs all flash, RNG, etc. operations to be done through its own APIs. Keeping the drivers tied to a particular subsystem could, for example, be to meet real-time and/or security requirements or because of hardware constraints.

Furthermore, we don't want to pollute the Zephyr config files, drivers, etc. with out-of-tree specific configuration and code.

Let's say that I have a subsystem `FOO` that when selected needs to use it's own driver `FLASH_DRIVER_FOO`. The problem arises when the user selects `CONFIG_FOO=y`, `FOO` has no way of saying that, e.g., `FLASH_DRIVER_NRF` is incompatible with it. 

I hope this cleared things a bit up, let me know if there is something I missed or didn't explain well enough.

Reply:
Let's go for CLOCK_CONTROL_HAS_OUT_OF_TREE_DRIVER

Reply:
... or maybe 

CLOCK_CONTROL_MUST_USE_OUT_OF_TREE_DRIVER

HAS implies that it is available to use, but in reality, the option indicates that the driver must be used.

Reply:
I agree on the semantics of HAS vs MUST, but I'd argue that naming consistency is more valuable. `<DRIVER>_HAS_OUT_OF_TREE_DRIVER` it is,

Reply:
But what about when in the future we HAVE an out-of-tree driver, but don't want to force it to be enabled?

Reply:
If you have a driver but it's not enforced, you'd add it to a choice selection. Then you can select driver at will and you shouldn't use this mechanism. I don't think that is an argument here.

By consistent naming I'm talking about using `HAS` to resolve `depends on` dependencies in a sane way. I would like to keep different names to a minimum, i.e., remove usage of `SUPPORTS` and the like. The I don't have to ask myself every time: "was it `<FOO>_HAS_<BAR>` or `<FOO>_SUPPORTS_<BAR>`, ..." 

That being said, _I do acknowledge that this case is different_ though, is it different enough to warrant its own convention?

Reply:
Yes.

Reply:
Changed